# My personal notes for Andrew Ng Machine Learning course

link:
https://www.youtube.com/playlist?list=PLZ9qNFMHZ-A4rycgrgOYma6zxF4BZGGPW

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import math

#### Transforming a for loop in a matrix multiplication

Note 1: if you want to convert a matrix (X) line in a matrix column use **X.reshape()**

Note 2: if you want to convert a vector into a matrix with one o more columns use **np.c_[np.ones((len(X),1)),X]**

See Example below

In [2]:
#Example
X=np.array([1,2,3,4])
print(f"{X}\n")
print(f"{X.reshape(4,1)}\n")
print(f"{np.c_[np.ones((len(X),2)),X]}")

del X

[1 2 3 4]

[[1]
 [2]
 [3]
 [4]]

[[1. 1. 1.]
 [1. 1. 2.]
 [1. 1. 3.]
 [1. 1. 4.]]


## Example

Imagine that you have a dataset Y with only one feature and you want to evaluate a funciton **f(x)=-40+0.25*x**.

We can do it as a for loop, but better with a matrix multiplication as follows:

$$Y=
\begin{pmatrix}
 2104\\
 1416\\
 1534\\
 852\\
\end{pmatrix}
$$

and
$$f(x)=-40+0.25\cdot~x$$

Solving through matrix multiplication we have:

$$f(Y)=
\begin{pmatrix}
 1 && 2104\\
 1 &&1416\\
 1 &&1534\\
 1 && 852\\
\end{pmatrix}
\times
\begin{pmatrix}
 -40\\
 0.25\\
\end{pmatrix}
=
\begin{pmatrix}
 486\\
 314\\
 343.5\\
 173\\
\end{pmatrix}
$$

In [3]:
Y = np.array([2104,1416,1534,852]) #dataset
#Transform the Y in a matrix 2x1 (rows x columns)
Y = np.c_[np.ones((len(Y),1)),Y]

fx = np.array([-40,0.25])

Answer = Y@fx #or Y.dot(fx)
print(Answer)

[486.  314.  343.5 173. ]


The same example above can be solve for multiple competing hypothesis f(x). Example:
$$Y=
\begin{pmatrix}
 2104\\
 1416\\
 1534\\
 852\\
\end{pmatrix}
$$

and
1. $$f(x)=-40+0.25\cdot~x$$
2. $$f(x)=200+0.1\cdot~x$$
3. $$f(x)=-150+0.4\cdot~x$$

Solving through matrix multiplication we have:

$$f(Y)=
\begin{pmatrix}
 1 && 2104\\
 1 &&1416\\
 1 &&1534\\
 1 && 852\\
\end{pmatrix}
\times
\begin{pmatrix}
 -40 && 200 && -150\\
 0.25 && 0.1 && 0.4\\
\end{pmatrix}
=
\begin{pmatrix}
 486 && 410 && 692\\
 314 && 342 && 416\\
 344 && 353 && 464\\
 173 && 285 && 191\\
\end{pmatrix}
$$

In [15]:
Y = np.array([2104,1416,1534,852]) #dataset
#Transform the Y in a matrix 2x1 (rows x columns)
Y = np.c_[np.ones((len(Y),1)),Y]

fx = np.array([[-40,200,-150],[0.25,0.1,0.4]])

Answer = Y@fx #or Y.dot(fx)
print(Answer)

[[486.  410.4 691.6]
 [314.  341.6 416.4]
 [343.5 353.4 463.6]
 [173.  285.2 190.8]]
